iati/validator.py

```diff
@@ -290,32 +356,29 @@ def _check_codes(dataset, codelist):
         base_xpath = mapping['xpath']
         condition = mapping['condition']
         split_xpath = base_xpath.split('/')
-        parent_el_xpath = '/'.join(split_xpath[:-1])
-        attr_name = split_xpath[-1:][0][1:]
+        last_xpath_section = split_xpath[-1]
 
-        if condition is None:
-            parent_el_xpath = parent_el_xpath + '[@' + attr_name + ']'
+        if last_xpath_section.startswith('@'):
+            located_codes = _extract_codes_from_attrib(dataset, split_xpath, condition)
+        elif last_xpath_section == 'text()':
+            located_codes = _extract_codes_from_element_text(dataset, split_xpath, condition)
         else:
-            parent_el_xpath = parent_el_xpath + '[' + condition + ' and @' + attr_name + ']'
-
-        # some nasty string manipulation to make the `//@xml:lang` mapping work
-        while not parent_el_xpath.startswith('//'):
-            parent_el_xpath = '/' + parent_el_xpath
-        if parent_el_xpath.startswith('//['):
-            parent_el_xpath = '//*[' + parent_el_xpath[3:]
-
-        parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
-
-        for parent in parents_to_check:
-            code = parent.attrib[attr_name]
+            raise ValueError('mapping path does not locate attribute value or element text')
 
+        for (code, line_number) in located_codes:  # `line_number` used via `locals()` # pylint: disable=unused-variable
             if code not in codelist.codes:
-                line_number = parent.sourceline  # used via `locals()` # pylint: disable=unused-variable
-
-                if codelist.complete:
-                    error = ValidationError('err-code-not-on-codelist', locals())
+                if last_xpath_section.startswith('@'):
+                    attr_name = last_xpath_section[1:]  # used via `locals()` # pylint: disable=unused-variable
+                    if codelist.complete:
+                        error = ValidationError('err-code-not-on-codelist', locals())
+                    else:
+                        error = ValidationError('warn-code-not-on-codelist', locals())
                 else:
-                    error = ValidationError('warn-code-not-on-codelist', locals())
+                    el_name = split_xpath[-2:-1][0]  # used via `locals()` # pylint: disable=unused-variable
```

Comment:
It takes me a while to get my head around this notation every time! I’d write this as:

```python
el_name = split_xpath[-2]
```

Reply:
More radical suggestion: I actually think this whole bit would be more legible if you scrapped `split_xpath`, used `rsplit('/', 1)` all over the place, and passed strings around instead of lists of strings. So like:

```python
parent_el_xpath, last_xpath_section = base_xpath.rsplit('/', 1)
```

But do feel free to ignore this suggestion, because this pull request looks good as it is!

Reply:
https://github.com/IATI/pyIATI/pull/229/commits/ec3f996030a7d4bfff6989b157ffc32773e9ede3

(I clearly haven't properly got my head around negative array indexing >.< - super useful feature that few languages I've used much have)

Reply:
With the `rsplit` version, it would require `_extract_codes_from_attrib()` and `_extract_codes_from_element_text()` to have different argument lists.

If these functions were public, that wouldn't be a great thing (lack of consistency). Since they're private, this isn't ideal, though is less of a problem.

It would, however, make the code a bit more DRY since `last_xpath_section` and `parent_el_xpath` would each only need calculating once - in `_check_codes()`. Hum.

Reply:
¯\_(ツ)_/¯

Reply:
0ec07bb265bac5b7ae37b08940c8a58d406c8cab

Reply:
Different args for those functions seems intuitively correct: one needs to know am attribute reference; the other doesn't.

I will stop peddling my reckons on this now, because I'm unsure so I trust whatever you decide!